### PR TITLE
Some test coverage (combinatorics and intfuncs)

### PR DESCRIPTION
**Pull request by hayd**
(Just a couple of low-hanging test files, regression tests.)

cc #9493

### CODE REVIEW

File: test/combinatorics.jl [hunk 1]
```diff
@@ -1,18 +1,35 @@
-@test factorial(7) == 5040
-@test factorial(7,3) == 7*6*5*4
 @test binomial(5,3) == 10
+@test binomial(2,1) == 2
+@test binomial(1,2) == 0
+@test binomial(-2,1) == -2
```
Comment (timholy) [inline test/combinatorics.jl]: I guess this makes sense, but I found it surprising. Add a comment to make sure someone doesn't delete it out of puzzlement later?

File: test/combinatorics.jl [hunk 2]
```diff
@@ -34,18 +51,16 @@ for n = 0:7, k = 1:factorial(n)
...
-# issue #6579
+@test factorial(7) == 5040
+@test factorial(7,3) == 7*6*5*4
 @test factorial(0) == 1
+@test_throws DomainError factorial(-1)
+
 @test factorial(int64(20)) == 2432902008176640000
+# issue #6579
 @test_throws OverflowError factorial(int64(21))
-@test_throws DomainError factorial(-1)
```
Comment (timholy) [inline test/combinatorics.jl]: Why is this deleted?

File: test/combinatorics.jl [hunk 3]
```diff
@@ -34,18 +51,16 @@ for n = 0:7, k = 1:factorial(n)
...
 @test typeof(factorial(int8(2))) == typeof(factorial(int8(1)))
 if Int === Int32
-@test factorial(int32(12)) === int32(479001600)
-@test_throws OverflowError factorial(int32(13))
+    @test factorial(int32(12)) === int32(479001600)
+    @test_throws OverflowError factorial(int32(13))
```
Comment (timholy) [inline test/combinatorics.jl]: whitespace here and above

**Comment by timholy** [comment]
Awesome!

File: test/combinatorics.jl (hunk 2 above)
Comment (hayd) [inline test/combinatorics.jl]: This is moved down (with the other factorial tests)

File: test/combinatorics.jl (hunk 3 above)
Comment (hayd) [inline test/combinatorics.jl]: this is indented since it's inside an if block. ?

**Comment by hayd** [comment]
Have an int32 failure:
```
exception on 2: ERROR: test failed: ("000004db" == "00000000000004db")
626 in expression: num2hex(1243) == "00000000000004db"
```
I did wonder if that might be an issue when I was writing it. Will post a fix, hopefully that's the only one.

File: test/combinatorics.jl (hunk 1 above)
Comment (hayd) [inline test/combinatorics.jl]: I was adding some edge cases this seems to agree with Wolfram Alpha!

File: test/intfuncs.jl [hunk 4]
```diff
@@ -14,3 +61,27 @@
...
+
+@test dec(121) == "121"
+
+@test hex(12) == "c"
+@test hex(-12, 3) == "-00c"
+@test num2hex(1243) == "00000000000004db"
+
+@test base(2, 5, 7) == "0000101"
+
+@test bits(1035) == "0000000000000000000000000000000000000000000000000000010000001011"
```
Comment (hayd) [inline test/intfuncs.jl]: will this be different on int32 also?

File: test/combinatorics.jl (hunk 3 above)
Comment (timholy) [inline test/combinatorics.jl]: Sorry, didn't notice the `if`.

**Comment by hayd** [comment]
@timholy Thanks for you comments, updated (and hopefully will pass appveyor now).

**Comment by hayd** [comment]
(rebased over https://github.com/JuliaLang/julia/commit/d877d63571982ccd6ecb2a0d370380456662e88d)

**Comment by timholy** [comment]
3/4 passes and one totally random-seeming failure. We must get to the bottom of these random failures, but I think adding more tests is a good step in that direction. So, merging.

**Comment by timholy** [comment]
Many thanks!

**Comment by tkelman** [comment]
may as well also backport these, with the exception of the base64 one which I don't think applies to release-0.3

**Comment by tkelman** [comment]
squashed and backported in 0582c44d5076151529c8de41dd55a27c32d7d077, and slightly tweaked to make pass on release-0.3

**Comment by tkelman** [comment]
Backporting the new tests revealed an interesting difference in behavior between 0.3 and 0.4 for `gcd` and `lcm` - 2508fcb8bd435287670deebbadef6e8b31a82ee6. Something for @juliabackports to keep in mind, hopefully obscure enough that it won't affect much.
